Add the fifth and final tutorial section. (#242)
square/workflow-kotlin: samples/tutorial/Tutorial5.md

# Step 5

_Unit and Integration Testing Workflows_

## Setup

To follow this tutorial, launch Android Studio and open this folder (`samples/tutorial`).

Start from the implementation of `tutorial-4-complete` if you're skipping ahead.

## Testing

`Workflow`s being easily testable was a design requirement. It is essential to building scalable, reliable software.

The `workflow-testing` library is provided to allow easy unit and integration testing. For this tutorial, we'll use the `kotlin-test` library to define tests and assertions, but feel free to use your favorite testing or assertion library instead – `workflow-testing` doesn't care.

## Unit Tests: `WorkflowAction`s

A `WorkflowAction`'s `apply` function is effectively a reducer. Given a current state and action, it returns a new state (and optionally an output). Because an `apply` function should almost always be a "pure" function, it is a great candidate for unit testing.

The `applyTo` extension function is provided to facilitate writing unit tests against actions.

> ### applyTo

Reviewer: Totally did not think of this method as a testing aid. Love it.

Reviewer: Why is it `@ExperimentalWorkflowApi`?

Author: I don't know. I'm down to remove that.

    assertEquals("New Title", state.todo.title)
    assertEquals("New Note", state.todo.note)
  }
```

## Unit Tests: Rendering

Testing actions is very useful for validating all of the state transitions of a workflow, but it is also beneficial to verify the logic in `render`. Since the `render` method uses a private implementation of a `RenderContext`, there is a `RenderTester` to facilitate testing.

### RenderTester

The `testRender` extension on `Workflow` provides an easy way to test the rendering of a workflow. It returns a `RenderTester` with a fluid API for describing test cases.

```kotlin
workflow.testRender(props = Props())
  .render { rendering ->
    assertEquals("expected text on rendering", rendering.text)
  }
```

It also provides a means to test that lambdas passed to screens cause the correct actions and state changes:

```kotlin
workflow.testRender(props = Props())
  .render { rendering ->
    assertEquals("expected text on rendering", rendering.text)
    rendering.updateText("updated")
  }
  .verifyActionResult { newState, output ->
    assertEquals(State(text = "updated"), newState)
  }
```

> The full API allows for expected workers and (child) workflows, as well as verification of resulting state and output:

Reviewer: ```suggestion
The full API allows for declaring expected workers and (child) workflows, as well as verification of resulting state and output:
```

```kotlin
workflow
  .testRender(
    props = Props(),
    initialState = State()
  )
  .expectWorkflow(
    workflowType = ChildWorkflow::class,
    rendering = ChildScreen(),
    output = WorkflowOutput(Closed)
  )
  .expectWorker(
    worker = TestWorker(),
    output = WorkflowOutput(Finished)
  )
  .render { rendering ->
    assertEquals("expected text on rendering", rendering.text)
  }
  .verifyActionResult { newState, output ->
    assertEquals(State(text = "updated"), newState)
    assertEquals(Completed, output?.value)
  }
```

### WelcomeWorkflow

Add tests for the rendering of the `WelcomeWorkflow`:

> ```swift

Reviewer: ```suggestion
```kotlin
```

Might want to `git grep` for that.

          assertEquals(2, rendering.size)
        }
        // Validate that the state was updated after the render pass with the output from the
        // TodoEditWorkflow.
        .verifyActionResult { newState, _ ->
          assertEquals(
              State(
                  todos = listOf(TodoModel(title = "Updated Title", note = "Updated Note")),
                  step = List
              ),
              newState
          )
        }
  }
}
```

## Integration Testing

The `RenderTester` allows easy "mocking" of child workflows and workers. However, this means that we are not exercising the full infrastructure (even though we could get a fairly high confidence from the tests). Sometimes, it may be worth putting together integration tests that test a full tree of Workflows. This lets us test integration with the non-workflow world as well, such as external reactive data sources that your workflows might be observing via Workers.

Add another test to `RootWorkflowTests`. We will use another test helper that spins up a real instance of the workflow runtime, the same runtime that `setContentWorkflow` uses.

### WorkflowTester

> When you create an Android app using Workflow, you will probably use `setContentWorkflow`, which starts a runtime to host your workflows, and wires it up to all the Android infrastructure. Under the hood, there's a lower-level API called `renderWorkflowIn` that runs the workflow runtime in a coroutine and exposes a `StateFlow` of renderings. When writing integration tests for workflows, you can use this API directly (maybe with a library like [Turbine](https://github.com/cashapp/turbine)), or you can use `workflow-testing`'s `WorkflowTester`. The `WorkflowTester` starts a workflow and lets you request renderings and outputs manually so you can write tests that interact with the runtime from the outside.

Reviewer: ```suggestion
When you create an Android app using Workflow, you will probably use `Activity.setContentWorkflow`, which starts a runtime to host your workflows, and wires it up to all the Android infrastructure. Under the hood, there's a lower-level API called `renderWorkflowIn` that runs the workflow runtime in a coroutine and exposes a `StateFlow` of renderings. When writing integration tests for workflows, you can use this API directly (maybe with a library like [Turbine](https://github.com/cashapp/turbine)), or you can use `workflow-testing`'s `WorkflowTester`. The `WorkflowTester` starts a workflow and lets you request renderings and outputs manually so you can write tests that interact with the runtime from the outside.
```

> This will be a "black box" test, as we can only test the behaviors from the rendering and will not be able to inspect the underlying states. This may be a useful test for validation when refactoring a tree of workflows to ensure they behave the same way.

Reviewer: ```suggestion
This will be an opaque test, as we can only test the behaviors from the rendering and will not be able to inspect the underlying states. This may be a useful test for validation when refactoring a tree of workflows to ensure they behave the same way.
```
